docs/developer/architecture/centralized-cli.rst

```diff
@@ -149,3 +149,203 @@ Cluster Service Administration
 +----------+---------------+-------------+----------------------------+
 | edit     | dex-config    | name        | open an editor             |
 +----------+---------------+-------------+----------------------------+
+
+Design Choices
+--------------
+
+Two distinct CLI:
+
+- a ``metalk8s`` kubectl plugin with subcommands to interact with Kubernetes
+  API, and Salt API if needed.
+- a ``metalk8sctl`` CLI with subcommands for action that need to interact with
+  the local machine, but may also interact with Kubernetes API and Salt API if
+  needed.
+
+
+``metalk8s`` kubectl plugin
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Language
+""""""""
+
+``Go`` is chosen as the language for kubectl plugin for the following reasons:
+
+- Great interactions with Kubernetes API.
+- Often used for operators and kubectl plugins
+  (`Sample CLI plugin <https://github.com/kubernetes/sample-cli-plugin>`_).
+- Portable as it's just a binary.
+- Simple deployment (no real requirements), just drop a binary in the PATH.
+
+Input and Output
+""""""""""""""""
+
+Each command should follow the kubectl style and standard as much as possible:
+
+- Command style:
+
+  .. code-block:: bash
+
+    kubectl metalk8s <action> <resource>
+
+- Interactive:
+
+  No interaction with the user, except when it's an ``edit`` command an editor
+  is opened (if needed).
+
+- Output style:
+
+  Default human-readable output (``<object> <action>ed``).
+
+  A ``--output``, ``-o`` option to change output format, at least support
+  for ``json`` and ``yaml`` (``jsonpath`` and ``go-template`` when it's
+  possible).
+
+  Internally each action result should be an "object"
+  (e.g.: single-level dictionary) containing several informations,
+  at least:
+
+  - name
+  - message
+  - result (True or False)
+  - an elapsed time (to know each action time)
+
+  A ``--verbose``, ``-v`` option to change log level verbosity (default output
+  to ``stderr``), using
+  `Kubernetes log library <https://github.com/kubernetes/klog>`_.
+
+
+SaltAPI interaction
+"""""""""""""""""""
+
+If the plugin needs to access Salt API a proxy should be configured by the
+plugin.
+
+And then plugin can access Salt API directly using the proxy
+``http://<server-ip>:<proxy-port>/api/v1/namespaces/kube-system/services/https:salt-master:api/proxy/``.
+
+This ``proxy-port`` should be configurable as it needs to be reachable by the
+plugin that can run from outside the MetalK8s cluster.
+
```

Comment:
Any chance we could consider something  like `pepperlib` from https://github.com/saltstack/pepper

Reply:
Nop, we do not need it, we will interact with API directly and we use Go
Anyway pepper is good to interact with salt-api directly as a CLI but it's not what we want here and even on the bootstrap we have access to the salt socket so we do not need pepper at all here.